File: demo/src/main/java/com/bytedance/scenedemo/MainScene.java
package com.bytedance.scenedemo;

import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import com.bytedance.scene.Scene;
import com.bytedance.scene.interfaces.ChildSceneLifecycleCallbacks;
import com.bytedance.scene.group.GroupScene;
import com.bytedance.scene.group.UserVisibleHintGroupScene;
import com.bytedance.scene.ui.GroupSceneUIUtility;
import com.bytedance.scenedemo.animation.AnimationListDemoScene;

import org.jetbrains.annotations.NotNull;

import java.util.LinkedHashMap;

/**
 * Created by JiangQi on 8/3/18.
 */
public class MainScene extends GroupScene {
    private ViewPager mViewPager;

    @NonNull
    @Override
    public ViewGroup onCreateView(@NotNull LayoutInflater inflater, @Nullable ViewGroup container,
                                  @Nullable Bundle savedInstanceState) {
        return (ViewGroup) inflater.inflate(R.layout.main_scene, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        this.mViewPager = view.findViewById(R.id.vp);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getNavigationScene().registerChildSceneLifecycleCallbacks(mChildSceneLifecycleCallbacks, true);

        if (getActivity() != null && getActivity().getWindow() != null) {
            Window window = getActivity().getWindow();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                window.getDecorView().setSystemUiVisibility(window.getDecorView().getSystemUiVisibility()
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getActivity().getWindow().setStatusBarColor(Color.TRANSPARENT);
            }
        }

        TabLayout tabLayout = findViewById(R.id.tablayout);
        tabLayout.setupWithViewPager(mViewPager);

        LinkedHashMap<String, UserVisibleHintGroupScene> list = new LinkedHashMap<>();

        list.put(getString(R.string.main_scene_title_navigation), new NavigationListDemo());
        list.put(getString(R.string.main_scene_title_part), new GroupListDemo());
        list.put(getString(R.string.main_scene_title_anim), new AnimationListDemoScene());
//        list.put(getString(R.string.main_scene_title_lifecycle), new LifeCycleListScene());
        list.put(getString(R.string.main_scene_title_extreme_case), new CaseListDemo());

        GroupSceneUIUtility.setupWithViewPager(this.mViewPager, this, list);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getNavigationScene().unregisterChildSceneLifecycleCallbacks(mChildSceneLifecycleCallbacks);
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    private ChildSceneLifecycleCallbacks mChildSceneLifecycleCallbacks = new ChildSceneLifecycleCallbacks() {
        @Override
        public void onSceneCreated(Scene scene, Bundle savedInstanceState) {
            log(scene.toString() + " onSceneCreated");
        }

        @Override
        public void onSceneActivityCreated(@NonNull Scene scene, @Nullable Bundle savedInstanceState) {
            log(scene.toString() + " onSceneActivityCreated");
        }

        @Override
        public void onSceneStarted(Scene scene) {
            log(scene.toString() + " onSceneStarted");
        }

        @Override
        public void onSceneResumed(Scene scene) {
            log(scene.toString() + " onSceneResumed");
        }

        @Override
        public void onSceneSaveInstanceState(Scene scene, @NonNull Bundle outState) {
            log(scene.toString() + " onSceneSaveInstanceState");
        }

        @Override
        public void onScenePaused(Scene scene) {
            log(scene.toString() + " onScenePaused");
        }

        @Override
        public void onSceneStopped(Scene scene) {
            log(scene.toString() + " onSceneStopped");
        }

        @Override
        public void onSceneViewDestroyed(@NonNull Scene scene) {
            log(scene.toString() + " onSceneViewDestroyed");
        }

        @Override
        public void onSceneDestroyed(Scene scene) {
            log(scene.toString() + " onSceneDestroyed");
        }
    };

    private static void log(String tag) {
        Log.e("Scene", tag);
    }
}

File: demo/src/main/java/com/bytedance/scenedemo/MainScene.kt
package com.bytedance.scenedemo

import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import com.bytedance.scene.Scene
import com.bytedance.scene.ui.template.NavigationViewScene
import com.bytedance.scenedemo.benchmark.BenchmarkSamples
import com.bytedance.scenedemo.other_library.OtherLibrarySamples
import com.bytedance.scenedemo.router.RouterSample


class MainScene : NavigationViewScene() {
    override fun getMenuResId(): Int {
        return R.menu.navigation_view
    }

    override fun getSceneMap(): LinkedHashMap<Int, Scene> {
        val map = LinkedHashMap<Int, Scene>()
        map[R.id.navigation] = NavigationListDemo()
        map[R.id.navigation_multi_stack] = MultiStackDemoScene()
        map[R.id.navigation_router] = RouterSample()
        map[R.id.group] = GroupSceneUsageSamples()
        map[R.id.dialog] = DialogSamples()
        map[R.id.animation] = AnimationListDemoScene()
        map[R.id.architecture_patterns] = PatternsSamples()
        map[R.id.activity_compatibility] = ActivityCompatibilitySamples()
        map[R.id.other_library] = OtherLibrarySamples()
        map[R.id.migrate] = MigrateSamples()
        map[R.id.compose] = ComposeSamples()
        map[R.id.benchmark] = BenchmarkSamples()
        map[R.id.extreme_case] = CaseListDemo()
        return map
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val window = requireActivity().window
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            window.decorView.systemUiVisibility = window.decorView.systemUiVisibility or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.statusBarColor = Color.TRANSPARENT
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            val lp: WindowManager.LayoutParams = window.attributes
            lp.layoutInDisplayCutoutMode =
                WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
            window.attributes = lp
        }
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onStop() {
        super.onStop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

}